amespahdbpythonsuite/parsers.py

```diff
@@ -0,0 +1,543 @@
+        """
+
+        schema = root.get('{http://www.w3.org/2001/XMLSchema-instance}' +
+                          'schemaLocation')
+        if schema:
+            _, uri = schema.split(' ', 1)
+            doc = ElementTree.parse(urllib.request.urlopen(uri))
+            xmlschema = ElementTree.XMLSchema(doc)
+            try:
+                xmlschema.assertValid(tree)
+            except Exception as e:
+                raise e
+            else:
+                self.validated = True
+
+    def _tree_to_dict(self, tree):
+        """Convert the tree to a dict of dicts.
+
+        Args:
+            tree: lxml tree of PAHdb file.
+
+        Returns:
+            database_dict: Dictionary, with the UIDs as keys,
+                containing the transitions, geometry data, as well as
+                UID metadata, references, comments.
+            info_dict: Dictionary containing general information about
+                the XML file, including filename, type, date, full
+                (bool), version, comment, and number of species.
+        """
+        # Advance to root and identify namespace.
+        event, root = next(tree)
+        namespace = self.determine_xml_namespace(root)
+
+        # Root dict attributes.
+        root_dict = root.attrib
+
+        # Form important dicts, to be returned.
+        database_dict = {}
+        info_dict = {}
+        first_run = True
+
+        # Iterate through the tree, clearing elements as we go.
+        for _, (event, elem) in enumerate(tree):
+
+            # Store the top XML comment.
+            if first_run:
+                if event == "end" and elem.tag == namespace + 'comment':
+                    pahdb_comment = elem.text
+                    first_run = False
+
+            # Extract UID data on a per-tag ('specie') basis.
+            if event == "end" and elem.tag == namespace + 'specie':
+                # Extract soft data.
+                uid, metadata, comments, reference = \
+                    self._extract_info(elem, namespace)
+
+                # Extract hard data.
+                args = elem, namespace
+                geometry_dict = self._extract_numeric(*args, 'geometry')
+                transitions_dict = self._extract_numeric(*args, 'transitions')
+                laboratory_dict = self._extract_numeric(*args, 'laboratory')
+
+                # Construct the UID-specific dictionary.
+                database_dict[uid] = {
+                    'metadata': metadata,
+                    'geometry': geometry_dict,
+                    'transitions': transitions_dict,
+                    'comments': comments,
+                    'reference': reference,
+                    'laboratory': laboratory_dict,
+                }
+
+                elem.clear()
+
+        # Path of XML file for saving in dict.
+        full_path = os.path.abspath(self.filename)
+
+        # Sanity check, no repeated UIDs.
+        db_keys = [key for key in database_dict]
```

Comment:
Is this not done by the schema-check; redundant here ...

Reply:
I don't think this is done in the schema check currently, unless this is a general requirement of the XML schema?